### PR TITLE
chore(*) inject ingress.kubernetes.io/service-upstream

### PR DESCRIPTION
**Pull request by nickolaev**
### Summary

Automatically inject `ingress.kubernetes.io/service-upstream` to all K8s Services, for the purposes of exposing Service Mesh service through the GW.
Previously users had to do this on their own.

### Documentation

- [ ] Link to the website [documentation PR](https://github.com/kumahq/kuma-website/pull/384)

### CODE REVIEW

**Comment by nickolaev** [comment]
@mergifyio update

File: pkg/plugins/runtime/k8s/controllers/service_controller.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+		}
+		return kube_ctrl.Result{}, errors.Wrapf(err, "unable to fetch Service %s", req.NamespacedName.Name)
+	}
+
+	injected, _, err := metadata.Annotations(namespace.Annotations).GetEnabled(metadata.KumaSidecarInjectionAnnotation)
+	if err != nil {
+		return kube_ctrl.Result{}, errors.Wrapf(err, "unable to check sidecar injection annotation on namespace %s", namespace.Name)
+	}
+	if !injected {
+		log.V(1).Info("not a Kuma service " + req.NamespacedName.String())
```
Comment (jakubdyszkiewicz) [inline pkg/plugins/runtime/k8s/controllers/service_controller.go:55]: nit `log.V(1).Info("service is not part of the mesh")`

File: pkg/plugins/runtime/k8s/controllers/service_controller.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+// ServiceReconciler reconciles a Service object
+type ServiceReconciler struct {
+	kube_client.Client
+	Log logr.Logger
+}
+
+// Reconcile is in charge of creating NetworkAttachmentDefinition if CNI enabled and namespace has label 'kuma.io/sidecar-injection: enabled'
+func (r *ServiceReconciler) Reconcile(req kube_ctrl.Request) (kube_ctrl.Result, error) {
+	log := r.Log.WithValues("service", req.Name)
```
Comment (jakubdyszkiewicz) [inline pkg/plugins/runtime/k8s/controllers/service_controller.go:31]: `req.Name` -> `req.NamespacedName` ?

File: pkg/plugins/runtime/k8s/controllers/service_controller.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+	if err != nil {
+		return kube_ctrl.Result{}, errors.Wrapf(err, "unable to check sidecar injection annotation on namespace %s", namespace.Name)
+	}
+	if !injected {
+		log.V(1).Info("not a Kuma service " + req.NamespacedName.String())
+		return kube_ctrl.Result{}, nil
+	}
+
+	svc.Annotations[metadata.IngressServiceUpstream] = metadata.AnnotationTrue
+
```
Comment (jakubdyszkiewicz) [inline pkg/plugins/runtime/k8s/controllers/service_controller.go:60]: how about log here:
```
log.Info("annotating service which is part of the mesh", "annotation", fmt.Sprintf("%s=%s", metadata.IngressServiceUpstream, metadata.AnnotationTrue)
```
?

**Review by jakubdyszkiewicz — APPROVED** [review]
Nice!

I see we are not covering the case when someone is injecting Kuma on the Pod, not the namespace. But since it's like ~95% cases, it should be fine 👌

**Pull request merged**